llumnix/queue/zmq_client.py

```diff
@@ -21,105 +21,127 @@
 import cloudpickle
 
 from llumnix.logging.logger import init_logger
+from llumnix.queue.queue_client_base import QueueClientBase
 from llumnix.server_info import ServerInfo
 
-from llumnix.queue.zmq_utils import (RPC_SUCCESS_STR, RPC_REQUEST_TYPE, RPCClientClosedError,
-                                     RPCUtilityRequest, RPCPutNoWaitQueueRequest, RPCPutNoWaitBatchQueueRequest,
-                                     get_open_zmq_ipc_path)
-from llumnix.constants import RPC_GET_DATA_TIMEOUT_MS, RPC_SOCKET_LIMIT_CUTOFF, RPC_ZMQ_HWM
+from llumnix.queue.zmq_utils import (
+    RPC_SUCCESS_STR,
+    RPC_REQUEST_TYPE,
+    RPCUtilityRequest,
+    RPCPutNoWaitQueueRequest,
+    RPCPutNoWaitBatchQueueRequest,
+    get_open_zmq_ipc_path,
+    get_zmq_socket_name,
+)
+from llumnix.constants import RPC_GET_DATA_TIMEOUT_MS, ZMQ_IO_THREADS
 from llumnix.metrics.timestamps import set_timestamp
 
 logger = init_logger(__name__)
 
 
-class ZmqClient:
+class SocketConnection:
+    def __init__(self, socket: zmq.asyncio.Socket):
+        self.socket: zmq.asyncio.Socket = socket
+        self.lock = asyncio.Lock()  # Ensures exclusive access to the socket across coroutines.
+
+    async def __aenter__(self):
+        await self.lock.acquire()
+        return self.socket
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        self.lock.release()
+
+    def close(self):
+        self.socket.close(linger=0)
+
+
+class ZmqSocketFactory:
+
+    def __init__(self, context: zmq.asyncio.Context):
+        self.context: zmq.asyncio.Context = context
+        self.socket_connections: Dict[str, SocketConnection] = {}
+
+    async def get_socket(self, ip, port) -> zmq.asyncio.Socket:
```

Comment:
this should not be a async function? Because when if dst_name in self.socket_connections, it's just a get dict operation.